src/graphlog-comparison/sapling-curved.txt

```diff
@@ -0,0 +1,225 @@
+        │ o  node 4
+        @ │  node 3
+        │ o  node 2
+        o  node 1
+        
+
+        o  node 3
+        │
+        ~
+
+        o  node 2
+        │
+        ~
+
+        @  node 1
+        │
+        ~
+        
+
+        o  node 4
+        │ o  node 3
+        o │  node 2
+        │ │
+        ~ │
+          │
+          o  node 1
+        
+
+        o  node 4
+        │ o  node 3
+        o │  node 2
+          │  more
+          │  text
+          o  node 1
+        
+
+        o  node 4
+        │ o  node 3
+        │ o  node 2
+        │ │
+        │ ~
+        │
+        o  node 1
+           more
+           text
+        
+
+        o  node 3
+        │ o  node 2
+        │ │  with
+        │ ~  long
+        │    description
+        o  node 1
+        
+
+        @  node 4
+        │ o  node 3
+        ├─╯
+        │ o  node 2
+        ├─╯
+        o  node 1
+        
+
+        o  node 10
+        │ o  node 9
+        │ │ o  node 8
+        o │ │  node 7
+        │ o │  node 6
+        │ │ o  node 5
+        o │ │  node 4
+        │ o │  node 3
+        ├─╯ │
+        │   o  node 2
+        ├───╯
+        o  node 1
+        
+
+        o  node 5
+        │ o  node 4
+        │ │ o  node 3
+        ├───╯
+        │ o  node 2
+        ├─╯
+        o  node 1
+        
+
+        o  node 7
+        │ o  node 6
+        │ │ o  node 5
+        │ │ │ o  node 4
+        ├─────╯
+        │ o │  node 3
+        ├─╯ │
+        │   o  node 2
+        ├───╯
+        o  node 1
+        
+
+        o  node 6
+        │ o  node 5
+        │ │ o  node 4
+        o │ │  node 3
+        ├───╯
+        │ o  node 2
+        ├─╯
+        o  node 1
+        
+
+        @        node 5
+        ├─┬─┬─╮  more
+        │ │ │ │  text
+        │ │ │ o  node 4
+        │ │ │ │
+        │ │ │ ~
+        │ │ │
+        │ │ o  node 3
+        │ │ │
+        │ │ ~
+        │ │
+        │ o  node 2
+        │ │
+        │ ~
+        │
+        o  node 1
+        │
+        ~
+        
+
+        o  node 8
+        │ o  node 7
+        │ │ o  node 6
+        │ │ │  with
+        │ │ │  some
+        │ │ │  more
+        │ │ │  lines
+        │ o │    node 5
+        │ ├───╮
+        │ o │ │  node 4
+        ├─╯ │ │
+        │   │ o  node 3
+        ├─────╯
+        │   o  node 2
+        ├───╯
+        o  node 1
+        
+
+        o  node 6
+        o      node 5
+        ├─┬─╮
+        │ │ o  node 4
+        │ o │  node 3
+        │ ├─╯
+        o │  node 2
+        ├─╯
+        o  node 1
+        
+
+        o  node 10
+        │ o  node 9
+        │ │ o  node 8
+        │ o │        node 7
+        │ ├───┬─┬─╮
+        │ o │ │ │ │  node 6
+        ├─╯ │ │ │ │
+        │   │ o │ │  node 5
+        ├─────╯ │ │
+        │   │   o │  node 4
+        ├───────╯ │
+        │   │     o  node 3
+        ├─────────╯
+        │   o  node 2
+        ├───╯
+        o  node 1
+        
+
+        @        node 1
+        ├─┬─┬─╮  with
+        │ │ │ │  many
```

Comment:
Not sure why this doesn't use dashed lines for the indirect edges. Might be a bug.

Reply:
I think this is because `~` currently means "an anonymous direct parent and we stop here - skip rendering its ancestors". `~` does not mean "a (distant) ancestor or root". A `~` might be one single commit without ancestors, in which case dashed lines are not appropriate.

Implementation wise,  `~` is used by [`Ancestor::Anonymous`](https://github.com/facebook/sapling/blob/5093c59f3f2fea4a3a4953596ea2b30e4d50292a/eden/scm/lib/dag/src/render/render.rs#L53). Only `Ancestor::Ancestor(_)` uses dashed lines.

Reply:
Yes, there are supposed to be `Ancestor` edges. Unless I have a bug.

Reply:
That seems strange. I think `Ancestor::Ancestor(_)` nodes occupy one line per node so it can have associated  commit messages. IIUC it's only possible to have multiple `~`s in one line by using `Ancestor::Anonymous`.

Reply:
Sorry, I commented on the wrong graph. It's the one below ([line 200](https://github.com/martinvonz/jj/pull/1048/files/12e17d4b689ba3723c681c6f1f8cdf40f58c7c70#diff-a97eeb0ef7cc0f5332817fe118d7d910566d6cb857380b2b9853de17248d174fR200) vs. [line 220](https://github.com/martinvonz/jj/pull/1048/files/12e17d4b689ba3723c681c6f1f8cdf40f58c7c70#diff-1ca0bda506fe1e2ffd7e9bc08dd53dca9fcd80a923e97c4d97c946703aa2a56fR220)). I think the "ancestor" part may just have length 0 in that drawing...

But let me double-check first that I'm calling this incorrectly. I'll get back to you.